File: internal/model/workflow_execution.go
package model

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/tuanvumaihuynh/roboflow/pkg/paging"
	"github.com/tuanvumaihuynh/roboflow/pkg/xerrors"
	"github.com/tuanvumaihuynh/roboflow/pkg/xsort"
)

type WorkflowExecutionStatus string

const (
	WorkflowExecutionStatusPending   WorkflowExecutionStatus = "PENDING"
	WorkflowExecutionStatusRunning   WorkflowExecutionStatus = "RUNNING"
	WorkflowExecutionStatusCompleted WorkflowExecutionStatus = "COMPLETED"
	WorkflowExecutionStatusFailed    WorkflowExecutionStatus = "FAILED"
	WorkflowExecutionStatusCanceled  WorkflowExecutionStatus = "CANCELED"
)

func (s WorkflowExecutionStatus) Validate() error {
	switch s {
	case WorkflowExecutionStatusPending:
	case WorkflowExecutionStatusRunning:
	case WorkflowExecutionStatusCompleted:
	case WorkflowExecutionStatusFailed:
	case WorkflowExecutionStatusCanceled:
	default:
		return xerrors.ThrowInvalidArgument(nil, "invalid workflow execution status")
	}
	return nil
}

type WorkflowExecution struct {
	ID          uuid.UUID
	WorkflowID  uuid.UUID
	Status      WorkflowExecutionStatus
	Env         map[string]string
	Definition  WorkflowDefinition
	CreatedAt   time.Time
	StartedAt   *time.Time
	CompletedAt *time.Time
	Steps       *[]Step
}

type WorkflowExecutionRepository interface {
	Get(ctx context.Context, id uuid.UUID) (WorkflowExecution, error)
	GetStatus(ctx context.Context, id uuid.UUID) (WorkflowExecutionStatus, error)
	List(ctx context.Context, workflowID uuid.UUID, p paging.Params, sorts []xsort.Sort) (*paging.List[WorkflowExecution], error)
	Create(ctx context.Context, workflowExecution WorkflowExecution) error
	Update(
		ctx context.Context,
		id uuid.UUID,
		fn func(wfe *WorkflowExecution) error,
	) error
	Delete(ctx context.Context, id uuid.UUID) error
}

File: ui/src/api/workflow-execution.ts
import type { SortPrefix } from '@/lib/sort'
import type { Paging } from '@/types/paging'
import type { WorkflowExecution } from '@/types/workflow-execution'
import http from '@/lib/http'

export type ListWorkflowExecutionSort = 'status' | 'created_at' | 'started_at' | 'completed_at'
export interface ListWorkflowExecutionParams {
  page?: number
  pageSize?: number
  sort?: SortPrefix<ListWorkflowExecutionSort>[]
}

const workflowExecutions = {
  list(workflowId: string, p: ListWorkflowExecutionParams): Promise<Paging<WorkflowExecution>> {
    return http.get(`/workflows/${workflowId}/executions`, { params: {
      page: p.page,
      pageSize: p.pageSize,
      sort: p.sort?.join(','),
    },
    })
  },
  get(id: string): Promise<WorkflowExecution> {
    return http.get(`/workflow-executions/${id}`)
  },
}

export default workflowExecutions
